pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+
+Providing more ways for users to find out about deprecated functionality
+can speed up the migration process. This PEP proposes to leverage static type
+checkers to communicate deprecations to users. Such checkers have a thorough
+semantic understanding of user code, enabling them to detect and report
+deprecations that a single ``grep`` invocation could not find. In addition, many type
+checkers integrate with IDEs, enabling users to see deprecation warnings
+right in their editors.
+
+Rationale
+=========
+
+At first glance, deprecations may not seem like a topic that type checkers should
+touch. After all, type checkers are concerned with checking whether code will
+work as is, not with potential future changes. However, the analysis that type
+checkers perform on code to find type errors is very similar to the analysis
+that would be needed to detect usage of many deprecations. Therefore, type
+checkers are well placed to find and report deprecations.
+
+Other languages already have similar functionality:
+
+* GCC supports a ``deprecated`` `attribute <https://gcc.gnu.org/onlinedocs/gcc-3.1.1/gcc/Type-Attributes.html>`__
+  on function declarations. This powers CPython's ``Py_DEPRECATED`` macro.
+* GraphQL `supports <https://spec.graphql.org/June2018/#sec-Field-Deprecation>`__
+  marking fields as ``@deprecated``.
+* Kotlin `supports <https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-deprecated/>`__
+  a ``Deprecated`` annotation.
+* Scala `supports <https://www.scala-lang.org/api/2.12.5/scala/deprecated.html>`__
+  an ``@deprecated`` annotation.
+* Swift `supports <https://docs.swift.org/swift-book/ReferenceManual/Attributes.html>`__
+  using the ``@available`` attribute to mark APIs as deprecated.
+* TypeScript `uses <https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html#deprecated>`__
+  the ``@deprecated`` JSDoc tag to issue a hint marking use of
+  deprecated functionality.
+
+Several users have requested support for such a feature:
+
+* `typing-sig thread <https://mail.python.org/archives/list/typing-sig@python.org/thread/E24WTMQUTGKPFKEXVCGGEFFMG7LDF3WT/>`__
+* `Pyright feature request <https://github.com/microsoft/pyright/discussions/2300>`__
+* `mypy feature request <https://github.com/python/mypy/issues/11439>`__
+
+
+Specification
+=============
+
+The new ``@typing.deprecated()`` decorator can be used on a class, function or method.
+With overloaded functions, the decorator may be applied to individual overloads,
+indicating that the particular overload is deprecated. The decorator may also be
+applied to the overload implementation function, indicating that the entire function
+is deprecated.
+
+The decorator takes a single argument of type ``str``, which is a message that should
+be shown by the type checker when it encounters a usage of the decorated object.
+The message must be a string literal or a string with a ``Literal`` type.
```

Comment:
If it's a string with a `Literal` type, shouldn't it be constrained to a single value? I wouldn't want to be a type checker having to implement something like
```py
message: typing.Literal["error", "warning"]

@typing.deprecated(message)
def f(x):
    ...
```

Reply:
Yes, it should be a single literal, I'll rephrase. I wanted to leave the door open for something like:

```python
MESSAGE: Final = "This is deprecated for reasons x, y, z. You should use Spam instead."

@deprecated(MESSAGE)
def func1(): ...

@deprecated(MESSAGE)
def func2(): ...
```